\documentclass{article}


\usepackage[T1]{fontenc}
\usepackage{natbib}
\usepackage{todonotes}
\usepackage{gensymb}
\usepackage{amssymb}
\usepackage{amsmath}
\usepackage{mathtools}
\usepackage[top=1in,bottom=1in,left=1in,right=1in]{geometry}
\usepackage{alltt}
\usepackage{framed}
\usepackage{xcolor}
\usepackage{comment}
\usepackage{graphicx}

\usepackage{xr}[]
\externaldocument{policar2019-tsne-embedding}
\usepackage{url}

\renewcommand{\thetable}{\Alph{table}}
\renewcommand{\thefigure}{\Alph{figure}}

\newcommand{\beq}{\begin{equation}}
\newcommand{\eeq}{\end{equation}}
\newcommand{\beqs}{\begin{equation}}
\newcommand{\eeqs}{\end{equation}}
\newcommand{\mx}[1]{\ensuremath{\boldsymbol{\mathrm{{#1}}}}}   % Simple matrix format.


% \newcommand{\review}[1]{\noindent{{\textbf{\textit{#1}}}}\vspace{3mm}}
% \definecolor{colframe}{rgb}{0.01,0.199,0.1}

\newcounter{rtaskno}
\newcommand{\rtask}[1]{\refstepcounter{rtaskno}\label{#1}}

\newcommand{\reviewc}[2]{\begin{leftbar}\noindent\rtask{#1}\ref{#1}. #2\end{leftbar}}
\newcommand{\review}[1]{\begin{leftbar}\noindent #1\end{leftbar}}
\newcommand{\answer}[1]{\noindent #1\\[0mm]}
\newcommand{\copyquote}[1]{\vspace{2mm} {\color{blue} #1}\vspace{2mm}}

\newlength{\leftbarwidth}
\setlength{\leftbarwidth}{3pt}
\newlength{\leftbarsep}
\setlength{\leftbarsep}{10pt}
\renewenvironment{leftbar}{%
    \def\FrameCommand{{\color{black}{\vrule width \leftbarwidth\relax\hspace {\leftbarsep}}}}%
    \MakeFramed {\advance \hsize -\width \FrameRestore }%
}{\endMakeFramed}


% \usepackage[colorlinks=True,citecolor=black]{hyperref}

\begin{document}


\noindent{\bf Poli\v{c}ar, Stra\v{z}ar \& Zupan: 
\\
Embedding to Reference t-SNE Space Addresses Batch Effects in Single-Cell Classification}

\vspace{5mm}
\noindent {\bf RESPONSE TO THE REVIEWERS}

\vspace{5mm}


\noindent Dear Editor and the Reviewers, \\
 
\vspace{3mm}

\noindent we thank for the comprehensive feedback that helped us to improve the
manuscript, sufficient time to implement the changes, and the permission to
submit a revised version. Below, we provide the detailed response to reviewers. Please note that the submission also includes a version of the manuscript with highlighted changes.

\vspace{3mm}

\noindent Kind regards,
\vspace{3mm}

\noindent Pavlin Poli\v{c}ar, 
Martin Stra\v{z}ar, 
Bla\v{z} Zupan \\ 
\noindent Ljubljana, June 18, 2021 

\vspace{10mm}


\subsection*{Reviewer \#1}

\reviewc{r1q1}{Most important is to add in theoretical part investigation about which dataset can be a reference dataset. It is necessary because it is crucial in estimation of level of trust worth. Some information about it is provided in testing part but only in form of general statement on specific case (based on some data sets), but it is necessary to have a procedure/test when given data set is a good reference data set.}

\answer{
We thank the reviewer for this comment. We have expanded our discussion with section 4.5, which highlights the need for a complete reference set. We also include an example of a failure case (Fig.6) where the reference data set violates this requirement.
}


\reviewc{r1q2}{Page 1
Bibliography:
positions 3 and 4 is not about PCA and MDS. Please provide appropriate citations for PCA and MDS.}

\answer{Thank you, we have fixed the error.}


\reviewc{r1q3}{
Page 6
How do you select L and $\sigma_{i,l}$?
}

\answer{
These are parameters of the t-SNE method, which was proposed in related work. We note that $L$ and the perplexities are user-specified parameters and overview the procedure for selection of $\sigma_{i,l}$ in the text (Sec. 3.1).
}


\reviewc{r1q4}{page 9
The "I(.)" in Eq. 11 is not defined.}

\answer{
We've added a short sentence, explaining that this is the indicator function. We moved this section to the appendix.
}


\reviewc{r1q5}{Please precise what do you mean by the right side of Eq. 12 (the meaning of brackets and "|").}

\answer{We have modified the text so that the equations in this section are clearer.}

\reviewc{r1q6}{Generally subsection 3.3 is not as clear as it should be. Please rewrite it.}

\answer{As suggested, we have improved the writing in Sec. 3.3. This background information on preprocessing of gene expression data is domain-specific and was hence moved to appendix.}


\reviewc{r1q7}{Figs 1-4
It would be much more readable when Figures would be bigger. Please provide subfigures vertically, not horizontally.}

\answer{Thank you for this suggestion. We have experimented with a different image layout and found that vertical alignment due to very large layout reduces readability of the text. We would kindly ask to leave this issue to the technical editor.}


\reviewc{r1q8}{GENERAL statement: provide better quality of all Figures (better resolution, now we see large pixels too early...)}

\answer{Thank you, all the figure are now rendered in higher quality.}


\reviewc{r1q9}{Fig 5
Please modify comparison presented in Fig 5.
Please provide 4 subfigures:
- t-SNE without PCA and without multiscale
- t-SNE with PCA and without multiscale
- t-SNE without PCA and with multiscale
- t-SNE with PCA and with multiscale
only in such case we can be sure what is real source of success.
It would be fine to see such comparison on (at least) two datasets. (!)}

\answer{
Multiscale kernels were introduced and studied by Kobak and Berens. With Fig. 5, we point that our pipeline can benefit from their approach as it produces a more coherent reference. Please note that our work focuses on adding new points into existing visualizations, regardless of how they were produced. Hence, we did not include further studies of creating reference visualizations as proposed by the reviewer.
}


\reviewc{r1q10}{Fig 6
Authors claim that in case of sub-figure c all genes may lead to under-clustering but the figures b and c are very similar in the number of clusters.}

\answer{
Globally, (b) and (c) are indeed similar, but there is a substantial difference in the structure of points describing neuronal cells. To clarify this point, we have addd ``In our example, the neuronal subclusters are more clearly defined in (b).'' to the caption of the figure.
}


\subsection*{Reviewer \#2}

\reviewc{r2q1}{Since the focus, from an application point of view, is on bioinformatics and specifically on batch effects, I think that a more detailed description of this phenomenon should be reported in the introduction. }

\answer{
We added a paragraph on batch effects to the section on related work.
}


\reviewc{r2q2}{The structure of the paper can significantly be improved. In particular:

- A section devoted to the description of related work should be introduced, reporting existing work on both the application and methodological viewpoints. In this respect, some descriptions currently reported in the experiments sections (see, for example, the description of scMap-cluster) should be moved in this section. Moreover, I actually can see a strong similarity with transfer learning methods, specifically with those focusing on domain adaptation. I suggest the authors to consider some relevant recent works in this field in their discussion (see, for example, the survey 10.1186/s40537-016-0043-6), also specifically focusing on bioinformatics (see, for example, the work 10.1093/bioinformatics/btz781).

- Related to the previous comment, some details related to existing approaches should be removed (or strongly simplified) from the section devoted to experiments, and moved to a new section devoted to related work. 

- Some parameter values reported in the methods section should be moved in the experiments, since they are part of the experimental setting.}

\answer{We thank the reviewer for this comment. We have restructured the paper accordingly. We have added a ``Related work'' section, where we describe batch-effects in more detail. We have also moved the description of scMap-Cluster from the ``Experiments'' section to this section. A relation with transfer learning is now addressed in Related work (second item in the second paragraph).
}

\reviewc{r2q3}{The current paper is an extension of a conference paper submitted to the conference Discovery Science 2019. The authors should explicitly provide in the text the actual novel contributions with respect to the conference version.}

\answer{
We added a paragraph to the end of the introduction explicitly listing our contributions of the paper. The differences from the conference version are mainly an improved experimental section, where we show that a naive placement of the data points leads to misleading conclusions, and that optimization produces more reasonable results. Additionally, we provide the classification evaluation where we compare the classification accuracy of our proposed approach and standard machine learning methods for scRNA-seq data.
}

\reviewc{r2q4}{The authors should better emphasize the contribution from a methodological viewpoint. Indeed, without a proper emphasis on this aspect, the paper seems to be the application of an existing tool (namely, openTSNE) to a new dataset. If the contribution from a methodological viewpoint is actually limited, the authors should still put some effort in emphasizing the advantages of this approach in this specific domain.}

\answer{
We have altered the wording in ``Availability and Implementation'' section to emphasize that the extension we describe in this paper was developed by us and later incorporated into \textsf{openTSNE}. As this is the only mention of \textsf{openTSNE} in the manuscript, we believe there can be no further confusion in this respect.
}


\reviewc{r2q5}{PCA-based initialization seems to help a lot. I would appreciate to see a comparison of the results with those achievable with the simple application of PCA with just 2 principal components.}

\answer{
PCA fails to produce cluster-informative visualizations of any high-dimensional data, as it focuses on variance instead of clustering structure. Comparison of these two methods have been already and extensively published (see, for instance, ``viSNE enables visualization of high dimensional single-cell data and reveals phenotypic heterogeneity of leukemia'' supplementary material).
}


\reviewc{r2q6}{At the beginning of Section 3.2, a quite long preprocessing pipeline is shown. Actually, I would like to understand the underline motivations behind that. Few additional sentences would help in this respect.}

\answer{
The single-cell processing pipeline we use is standard, and we have provided references to original proposals of the methods. We include the description of preprocessing pipeline in the paper for the reasons of reproducibility. As data preprocessing is not the focus of our work, we have moved the description to the Appendix.}


\subsection*{Reviewer \#3}

\reviewc{r3q1}{The research heavily relies on the ref. [15]. Although the authors claimed that they provide both practical and theoretical grounded implementation, the contribution in terms of theoretical justification is not convincing enough. The equations, given in Section 2, closely correlate with those provided in [15], except for two small modifications - replacing x and y with v and w. }

\answer{
Our work extends the original t-SNE approach and uses a bag of tricks by Kobak and Berens. The reviewer is right: our formulation of the loss function for new data points is (intentionally) very similar to that of original t-SNE's. Our main contributions are the first demonstration that such an approach can work, produces excellent results, and even removes the batch effects. We expose these novelties in the new paragraph at the end of the introduction and highlight the importance of steps in our approach in the last paragraph of Sec.3.2.
}


\reviewc{r3q2}{From the Abstract, it is not immediately clear what the reference and secondary datasets are. The authors explained it later in the main body, but it seems reasonable to introduce the terms earlier to keep readers interested. Further, the authors claim that they show "the utility of the proposed approach by analyzing 6 recently published ... datasets … with up to tens of thousands of cells and thousands of genes ..."  and so on. Mentioning more specific quantitative results with regard to particular datasets, in comparison with the known results, would be desirable. }

\answer{
We have reworded the abstract to indicate that the secondary data set contains new, unseen points. We agree with a reviewer that quantifying the results is important. Hence, the manuscript includes a comparison with other techniques (Table 2). We, however, refrain from reporting the accuracy estimates in the abstract, as the focus of our work is a visualization and not optimization of classification accuracy.
}


\reviewc{r3q3}{The Results and Discussion section is difficult to follow. Please, if possible, consider splitting it in subsections.}

\answer{
We thank the reviewer for this comment. We have split up the ``Results and Discussion'' section into several subsections.
}


\reviewc{r3q4}{Consider replacing the legends in Fig. 1 with the names or descriptions of the datasets to increase the readability. }

\answer{
We want to thank the reviewer for this comment. Both datasets contain cells as indicated in the panel title (brain, pancreas) and use a similar approach to gather the data. The main difference is in the group that performed the experiments and published the data. Hence, we use the first authors of the publications to distinguish between two batches of data.
}


\reviewc{r3q5}{Please extend Section 3.5, which includes only two sentences, with more details of implementation or consider moving this text into Section 3.4 (possibly at the beginning or at the end).}

\answer{We agree that a particular section is not the most appropriate location for this kind of information. We have moved the note on implementation to a more suitable place in the note before the acknowledgments.}


\reviewc{r3q6}{Please update the preprint [15] with the published manuscript.}

\answer{The thank the reviewer for the reminder. We have updated the citation accordingly.}

\end{document}
